pkg/proxy/metaproxier/meta_proxier.go

```diff
@@ -79,11 +103,88 @@ func (proxier *metaProxier) OnServiceUpdate(oldService, service *v1.Service) {
 	if utilproxy.ShouldSkipService(service) {
 		return
 	}
-	// IPFamily is immutable, hence we only need to check on the new service
-	if utilnet.IsIPv6String(service.Spec.ClusterIP) {
-		proxier.ipv6Proxier.OnServiceUpdate(oldService, service)
-	} else {
-		proxier.ipv4Proxier.OnServiceUpdate(oldService, service)
+	// case zero: this allows skew between new proxy and old apiserver
+	if len(service.Spec.IPFamilies) == 0 {
+		actual := proxier.getProxyByClusterIP(oldService)
+		actual.OnServiceUpdate(oldService, service)
+		return
+	}
+
+	// case one: something has changed, but not families
+	// call update on all families the service carries.
+	if len(oldService.Spec.IPFamilies) == len(service.Spec.IPFamilies) {
+		for _, ipFamily := range service.Spec.IPFamilies {
+			actual := proxier.getProxierByIPFamily(ipFamily)
+			actual.OnServiceUpdate(oldService, service)
+		}
+
+		klog.V(4).Infof("service %s/%s has been updated but no ip family change detected", service.Namespace, service.Name)
+		return
+	}
+	// while apiserver does not allow changing primary ipfamily
+	// we use the below approach to stay on the safe side.
+	// note: in all cases, we check all families just
+	// in case the service moved from ExternalName => ClusterIP
+	// or the other way around
+
+	// case two: service was upgraded (+1 ipFamily)
+	// call add for new family
+	// call update for existing family
+	// note: Service might have been upgraded and
+	// had port/toplogy keys etc  changed.
+	if len(service.Spec.IPFamilies) > len(oldService.Spec.IPFamilies) {
+		found := false
+		for _, newSvcIPFamily := range service.Spec.IPFamilies {
+			for _, existingSvcIPFamily := range oldService.Spec.IPFamilies {
+				if newSvcIPFamily == existingSvcIPFamily {
+					found = true
+					break
+				}
+			}
+
+			actual := proxier.getProxierByIPFamily(newSvcIPFamily)
+			if found {
+				actual.OnServiceUpdate(oldService, service)
+			} else {
+				// TODO (khenidak) once we do uber
+				// port locking, we will need to release before
+				// calling the actual proxier here. since
+				// actual proxier also attempts to listen to port
+				klog.V(4).Infof("service %s/%s has been updated and ipfamily %v was added", service.Namespace, service.Name, newSvcIPFamily)
+				actual.OnServiceAdd(service)
```

Comment:
seems that everything calls to OnServiceUpdate() inside the proxiers, you can simplify this logic just checkin if you have to pass the oldService or nil to the method

https://github.com/kubernetes/kubernetes/blob/f3a8666e4ab96c08284d2695c64e0e4428c5460b/pkg/proxy/ipvs/proxier.go#L854-L869

Reply:
The idea is to keep most of existing logic as is and wrap it on meta proxier. that way actuals can change without having to change anything in meta proxy

Reply:
that makes sense, thanks